source/magpie/group.d

```diff
@@ -0,0 +1,154 @@
+    GrpType data;
+
+    /// Index for group
+    Index grpIndex;
+
+public:
+    /++
+    int getGroupPosition(string[] grpTitles)
+    Description: Get position of a particular group in the sruct
+    @params: grpTitles - the group you wnant to search for
+    +/
+    int getGroupPosition(string[] grpTitles)
+    {
+        foreach(i, ele; groups)
+            if(grpTitles == ele)
+                return cast(int)i;
+        
+        return -1;
+    }
+
+    /++
+    void createGroup(T)(DataFrame!T df, int[] levels)
+    Description: Funtion to create a groupBy object from a DataFrame
+    @parmas: df - DataFrame on which groupBy is operated
+    @params: levels - levels to consider for grouping
+    +/
+    void createGroup(int[] dataLevels, T...)(DataFrame!T df, int[] indexLevels)
+    {
+        import std.algorithm: reduce, max, min, sort;
+        assert(indexLevels.reduce!max < df.indx.row.index.length);
+        assert(indexLevels.reduce!min > -1);
+        assert(dataLevels.reduce!max < df.RowType.length);
+        assert(dataLevels.reduce!min > -1);
+
+        string[][] levels;
+        levels.length = df.rows;
+        foreach(i; 0 .. df.rows)
+        {
+            levels[i].length = indexLevels.length + dataLevels.length;
+            foreach(j, ele; indexLevels)
+            {
+                if(df.indx.row.index[ele].length == 0)
+                {
+                    import std.conv: to;
+                    levels[i][j] = to!string(df.indx.row.codes[ele][i]);
+                }
+                else
+                    levels[i][j] = df.indx.row.index[ele][df.indx.row.codes[ele][i]];
+            }
+
+            static foreach(j, ele; dataLevels)
+            {
+                static if(is(df.RowType[ele] == string))
+                    levels[i][indexLevels.length + j] = df.data[ele][i];
+                else
+                {
+                    import std.conv: to;
+                    levels[i][indexLevels.length + j] = to!string(df.data[ele][i]);
+                }
+            }
+        }
+        
+        // This si required as dropper!(dataLevels, df.data) won't work
+        auto dropped = df.drop!(1, dataLevels);
+
+        data = dropped.data;
+        grpIndex.column = dropped.indx.column;
+        grpIndex.row.titles = dropper(indexLevels, df.indx.row.titles);
+        grpIndex.row.index = dropper(indexLevels, df.indx.row.index);
+
+        import std.range: zip;
+        int[] codes = vectorize(levels);
+        int[][] rcodes = transposed(dropper(indexLevels, df.indx.row.codes));
+        // Simultaneously arranging all the relavent fields using a zip as displayed in the docs
+        auto arrange = zip(levels, codes[1 .. $], data, rcodes).sort!((a, b) => a[1] < b[1]);
+
+        grpIndex.row.codes = transposed(rcodes);
+
+        elementCountTill.length = codes[0];
+        groups.length = codes[0];
+        groups[0] = levels[0];
+        elementCountTill[0] = 0;
+        int uele = 0, count = 0;
```

Comment:
all variables are default initialised to their `.init` property, for `int` this is `0`.

Reply:
:sweat_smile: Addressed in 6517b4a